### ISSUE TITLE
FIFO Queues with delays

### ISSUE THREAD

**Issue by davidrichey** [phstc/shoryuken#643]
When using FIFO queues & configuring a delay, there will only be 1 message processed per delay period. As an example I have 2 messages in the SQS queue. With a 5 second shoryuken `delay` & the job taking 2 seconds to complete, the process to fetch will wait the 5 seconds to go and fetch the next message from the queue. This proposal would allow for the fetching of the FIFO queue which is paused after a worker has completed the work so the delay would kick in only after fetching with 0 messages being returned.

Configuration:
```
concurrency: 5
delay: 5
queues:
  - ordered.fifo
```

shoryuken.rb
```
Shoryuken.configure_server do |config|
  config.on(:dispatch) do
    # called every time it tries to fetch messages from SQS
    puts "#{Time.now}: Fetching Message..."
  end
end
```

Job.rb
```
class Job < ApplicationJob
  self.queue_adapter = :shoryuken
  include Shoryuken::Worker
  queue_as "ordered.fifo"

  def perform(message)
    puts message
    sleep 2
  end
end
```

Logs from running shoryuken with a 5 second delay:

```
2021-01-18T16:39:40Z 6597 TID-owbyxurl8 INFO: Starting
2021-01-18 11:39:40 -0500: Fetching Message...
2021-01-18T16:39:40Z 6597 TID-owbzp8mn0 ActiveJob/Job/development-ordered.fifo/5142b4e9-e146-4e26-b20a-bc44522449d1 INFO: started at 2021-01-18 11:39:40 -0500
2021-01-18 11:39:40 -0500: Fetching Message...
2021-01-18T16:39:42Z 6597 TID-owbzp8mn0 ActiveJob/Job/development-ordered.fifo/5142b4e9-e146-4e26-b20a-bc44522449d1 INFO: completed in: 2127.785 ms
2021-01-18 11:39:45 -0500: Fetching Message...
2021-01-18T16:39:45Z 6597 TID-owbzp8mn0 ActiveJob/Job/development-ordered.fifo/7ec83d58-3eef-485e-b0f7-a349df522117 INFO: started at 2021-01-18 11:39:45 -0500
2021-01-18 11:39:45 -0500: Fetching Message...
2021-01-18T16:39:47Z 6597 TID-owbzp8mn0 ActiveJob/Job/development-ordered.fifo/7ec83d58-3eef-485e-b0f7-a349df522117 INFO: completed in: 2033.8250000000003 ms
```

:pushpin: #644

**Comment by davidrichey** [comment]
I have added a draft PR (#644) to start the conversation on implementation

**Comment by cjlarose** [comment]
Just in case you hadn't seen it, #628 is an open PR that also addresses the throughput problem for FIFO queues.

The new PR #644 looks like it addresses [@phstc's feedback](https://github.com/phstc/shoryuken/pull/628#pullrequestreview-501558587) on #628 since it doesn't hold onto messages in memory (and thereby allowing the visibility timeout clock to start ticking prematurely).

The basic idea of unpausing the FIFO queue whenever we finish work on a message seems solid, but I have a few questions to get the conversation started:

1. Do you think it is possible to implement something similar without chaining the `polling_strategy` through the middleware stack? As implemented, I think doing this breaks existing user-defined middleware since it changes the middleware interface in a backwards-incompatible way (users have their own middlewares that they load into the middleware chain).
2. How does this interact with messages that have different message group IDs? #644 doesn't seem to handle those specifically, which I think is okay, but I just wanted to make sure. By contrast #628 seems to have attempted to make it so that shoryuken can process multiple messages from a FIFO queue concurrently (as long they have different message group IDs). #644 makes no attempt to do this if I understand it correctly, but I think that's okay--the new behavior is no worse than the current behavior--it still restricts shoryuken to processing one messages in a FIFO queue at a time.

**Comment by cjlarose** [comment]
Curious on your thoughts, too, @cstyles. Seems like the basic idea of #644 is less ambitious than your own #628, but would definitely improve throughput on FIFO queues.

**Comment by davidrichey** [comment]
Thanks @cjlarose. I did read over #628 but didn't see it until this morning. 

> The new PR #644 looks like it addresses @phstc's feedback on #628 since it doesn't hold onto messages in memory (and thereby allowing the visibility timeout clock to start ticking prematurely).

Agreed 👍  

> Do you think it is possible to implement something similar without chaining the polling_strategy through the middleware stack? As implemented, I think doing this breaks existing user-defined middleware...

Noted. I'll have to dig back in on this. I saw feedback on [codeclimate](https://codeclimate.com/github/phstc/shoryuken/pull/644) on those methods anyways. Any tips on this would be appreciated as I've just started digging into this codebase

> How does this interact with messages that have different message group IDs?

The attached PR would not change behavior. I would have do a lot more testing/playground to attempt to implement something like that. That would go beyond our need at the moment.

**Comment by cjlarose** [comment]
>  Any tips on this would be appreciated as I've just started digging into this codebase

My gut feeling is that this logic can be implemented in the Manager without going through the middleware stack. After a job has completed the [`processor_done`][0] method is invoked. There, we have a reference to the `polling_strategy` in an instance variable, so that's the place where I think we can unpause the queue.

[0]: https://github.com/phstc/shoryuken/blob/1a261d66a678a41266e389642c54bbf07819ee21/lib/shoryuken/manager.rb#L60-L62

> The attached PR would not change behavior. I would have do a lot more testing/playground to attempt to implement something like that. That would go beyond our need at the moment.

That makes sense. A change that improves FIFO throughput without changing the behavior around message group IDs would still of course be welcome.

**Comment by davidrichey** [comment]
Awesome, thanks @cjlarose. Took another look with that information from above. I was able to access the polling strategy with the Manager 👍 . Pushed the updates to that PR.

**Comment by cjlarose** [comment]
@davidrichey Thanks for updating the PR. New changes look good. I'll take a stab at doing some testing on it myself and give @cstyles a chance to chime in, too.

**Comment by cstyles** [comment]
Sorry it took me a while to get to this! I'll just dump some thoughts I had while catching up on all this.

I like the basic idea here. However, it's worth mentioning that this bit...

> When using FIFO queues & configuring a delay, there will only be 1 message processed per delay period

...is only true when `concurrency` is greater than 1. In production at my company, we run shoryuken for one of our apps with concurrency set to 1 so that this conditional...

https://github.com/phstc/shoryuken/blob/1a261d66a678a41266e389642c54bbf07819ee21/lib/shoryuken/manager.rb#L37-L39

...always evaluates to `true` after fetching & dispatching a message in a FIFO queue. This causes shoryuken to busy-wait until the in-flight job is finished before fetching another message. The delay is only incurred when there are available processors and we try to fetch another message (to which AWS will return 0 messages because the previous message hasn't been processed yet).

Obviously setting `concurrency` to 1 isn't an acceptable solution for all use cases so I do think this proposal (or something like it) would be nice.

Incidentally, I just learned that shoryuken allows configuring `concurrency` per-queue via [processing groups](https://github.com/phstc/shoryuken/wiki/Processing-Groups). So it seems the current "recommended" way to avoid unnecessary delays is to ensure your FIFO queues all have `concurrency` set to 1. Which makes me wonder: are there any situations where we actually want a FIFO queue worker to have `concurrency` greater than 1? If not, can we make shoryuken automatically set `concurrency` to 1 for all FIFO queues unless otherwise specified?

**Comment by davidrichey** [comment]
Thanks @cstyles! I have not tried the concurrency of one to have the manager busy-wait.

> It occurred to me immediately after commenting that if the FIFO queue contains messages with different message group IDs, it could be useful to have multiple processors

This is exactly our use case. We are building out a queueing system to be able to process many messages all within a single FIFO queue and hoping to separate the processing by using message_group_id. With this use case we wouldn't be able to set the concurrency to 1.

The goal is to still be able to use the delay so we limit our fetching messages while still processing as fast as possible when the queue is full.

This implementation was the only way I could see for our use case, any other suggestions?

**Comment by cstyles** [comment]
I spoke with @cjlarose offline and we came to the conclusion that your PR is the best option.

For posterity, another solution we discussed is to define a separate [processing group](https://github.com/phstc/shoryuken/wiki/Processing-Groups) for each FIFO queue and set its `concurrency` to 1. That way, we get the busy-waiting behavior, plus queues won't block other queues like they would if the global `concurrency` setting was 1. However, this doesn't allow for concurrent execution of messages with different message group IDs in the same queue so it's not a great solution for your use case. It also requires a bunch of configuration to get it working correctly. By contrast, users of shoryuken who upgrade would reap the benefits of the unpause solution without any action on their part.

One potential downside of the unpause idea is that it will likely make more calls to SQS than the above idea. For example:
1. We fetch a message from SQS and start processing it.
2. We try to fetch another message. SQS tells us there are no messages available because the previous message is still being processed. We pause the queue.
3. The message is finished processing and we unpause the queue.
4. Go to step 1.

So in the worst case we end up making two SQS calls for each message. However, that second fetch (the one that incurs the delay) could return messages of a different group ID so it's not useless. Also it's no worse than the current behavior when `concurrency` is greater than 1.

All that said, I'm on board with the proposed change. I haven't looked at the actual code yet but conceptually it sounds good to me. 👍

**Comment by cjlarose** [comment]
Yeah, I think one way to think about the proposed change is by considering each possible relevant configuration a user might have.

1. For users that have a FIFO queue in a processing group with `concurrency: 1`, there's no effect. No new workers are available to pick up work, so a fetch attempt is never made unnecessarily and the queue is never paused unnecessarily.
2. For users that have a FIFO queue in a processing group with `concurrency > 1`, they'll get better throughput at the cost of more API calls, regardless of whether or not they have messages in the queue with distinct message group IDs.

I think one takeaway from this is that if the user has a FIFO queue where they have only a single distinct message group ID, then they should consider putting it in a processing group with concurrency set to 1. That'll avoid unnecessary fetches from the queue where we know that no messages could possibly be available. The cost, of course, is that you're basically "reserving" a thread for processing messages on that queue. That said, this recommendation is kinda orthogonal to the proposed PR--folks should consider doing this in the current version of Shoryuken if they're in this situation already.
